src/main.rs

```diff
@@ -394,7 +408,61 @@ async fn run() -> Result<()> {
         PathBuf::from(&src)
     };
 
-    let mut choices = Vec::new();
+    // Let's first find all subdirectories that are compatible with our supported recipes.
+    let root_src_dir = src_dir.clone();
+    let mut open_subdirs = vec![(0, src_dir)];
+    let mut sourceroot_candidates = vec![];
+    while !open_subdirs.is_empty() {
+        // If we call this, we have a non-empty vector.
+        let (depth, target_dir) = open_subdirs.pop().unwrap();
+        let subdirs = std::fs::read_dir(&target_dir)
+            .context(format!("failed to list subdirectories of {}", target_dir.to_string_lossy()))?;
+
+        for maybe_subdir in subdirs {
+            let subdir = maybe_subdir.context(
+                format!("unexpected io error while reading a subdirectory of {}", target_dir.to_string_lossy())
+            )?;
+
+            // We want only subdirectories.
+            if !subdir.path().is_dir() {
+                continue;
+            }
+
+            if is_target_for_packaging(subdir.path()) {
+                sourceroot_candidates.push(subdir.path());
+            }
+
+            // Recurse into that subdir.
+            if depth + 1 <= 3 {
+                open_subdirs.push((depth + 1, subdir.path()));
+            }
+        }
+    }
+
+    editor.set_helper(Some(Prompter::List(
+                sourceroot_candidates
+                .iter().map(|sr| sr
+                    .strip_prefix(&root_src_dir)
+                    .expect("Failed to strip the root source directory prefix")
+                    .to_string_lossy()
+                    .to_string())
+                .collect()
+    )));
+    let choice = editor.readline(&prompt("Select which subdirectory should we use?"))?;
+    //let Some(Prompter::List(_choices)) = editor.helper_mut() else {
+    //    unreachable!();
+    //};
+    let src_dir = choice
+        .parse()
+        .ok()
+        .and_then(|i: usize| sourceroot_candidates.get(i))
+        .unwrap_or_else(|| &sourceroot_candidates[0]);
```

Comment:
we should still allow packages to opt out of sourceRoot even when there are candidates

Reply:
hmmm why?

Reply:
workspaces?

Reply:
e.g. there can be a rust binary at root that also have a library and a python binding of the library in a subdirectory, but we might want the rust binary instead of the python binding

Reply:
We only have to make it so that sourceRoot candidate include the root src dir, right?

Reply:
actually why
```
    let src_dir = choice
        .parse()
        .ok()
        .and_then(|i: usize| sourceroot_candidates.get(i))
        .unwrap_or_else(|| &sourceroot_candidates[0]);
    let source_root = if src_dir != &root_src_dir {
        Some(src_dir)
    } else {
        None
    };
```

is not enough?

Reply:
if your choice is the rootdir, then you don't get a sourceRoot

Reply:
that would also work, but we would need extra logic to be able to omit the `sourceRoot =` part

Reply:
> actually why <...> is not enough?

I haven't check since the current HEAD is not building for me, but I think `source_candidates` never contains `root_src_dir`

Reply:
yes, fixing